test/common/parallel_tests.js

```diff
@@ -57,101 +77,168 @@ function parallelTests(tag, suite, section) {
+
+		const completedFiles = {};
+
+		const next = () => {
+			const testFile = allFiles.splice(0, 1);
+			spawnParallelTest(testFile, mochaArguments)
+				.then(testFile => {
+					completedFiles[testFile] = 'done';
+					if (
+						allFiles.length === 0 &&
+						Object.keys(completedFiles).length === allFilesLength
+					) {
+						console.info('All parallel tests finished successfully.');
+						return resolve('All parallel tests finished successfully.');
+					}
+
+					if (allFiles.length > 0) {
+						return next();
+					}
+				})
+				.catch(err => {
+					console.error(`Parallel test failed: ${testFile}`);
+					return reject(err);
+				});
 		};
 
+		for (
+			let i = 0, limit = Math.min(allFilesLength, maxParallelism);
+			i < limit;
+			i += 1
+		) {
+			next();
+		}
+	});
+};
+
+const runStandaloneTests = (suiteFolder, mochaArguments) => {
+	return new Promise((resolve, reject) => {
+		const child = executeIstanbul(suiteFolder, mochaArguments);
 		child.on('close', code => {
 			if (code === 0) {
-				console.info('Test finished successfully:', test);
-				delete parallelTestsRunning[child.pid];
-
-				if (filepaths.length) {
-					spawnTest(filepaths.shift());
-				}
-				if (Object.keys(parallelTestsRunning).length === 0) {
-					return console.info('All tests finished successfully.');
-				}
-				return;
+				console.info('All standalone tests finished successfully.');
+				return resolve();
 			}
 
-			console.info('Test failed:', test);
-			cleanupRunningTests();
-			process.exit(code);
+			console.error('Standalone tests failed:', suiteFolder);
+			child.kill('SIGTERM');
+			reject(code);
+			return process.exit(code);
 		});
 
 		child.on('error', err => {
 			console.error(err);
-			cleanupRunningTests();
-			process.exit();
+			child.kill('SIGTERM');
+			reject(err);
+			return process.exit();
 		});
-	};
+	});
+};
 
-	initFilepaths.forEach(spawnTest);
+function executeTests(tag, suite, section) {
+	return new Promise(async (resolve, reject) => {
+		console.info('Executing tests with following configuration:', {
+			tag,
+			suite,
+			section,
+		});
+		const suiteFolder = getSuiteFolder(suite, section);
+		const mochaArguments = getMochaArguments(tag);
+
+		if (tag !== 'standalone') {
+			try {
+				const result = await runParallelTests(suiteFolder, mochaArguments);
+				return resolve(result);
+			} catch (err) {
+				console.error('Parallel tests failed!', err);
+				return reject(err);
+			}
+		} else {
+			try {
+				const result = await runStandaloneTests(suiteFolder, mochaArguments);
+				return resolve(result);
+			} catch (err) {
+				console.error('Standalone tests failed!', err);
+				return reject(err);
+			}
+		}
+	});
 }
 
-parallelTests(process.argv[2], process.argv[3], process.argv[4]);
+(async () => {
+	await executeTests(process.argv[2], process.argv[3], process.argv[4]);
+	if (process.argv[2] !== 'standalone') {
+		await executeTests('standalone', process.argv[3], process.argv[4]);
```

Comment:
In case of extensive, we will run `standalone` tests twice?

Reply:
We shouldn't.

```js
case 'extensive':
			mochaArguments.push('--', '--grep', '@unstable|@standalone', '--invert');
			break;
```

We are skipping standalone tests with `--invert` in extensive tests too.